dali/pipeline/data/sample_view.h

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_DATA_SAMPLE_VIEW_H_
+#define DALI_PIPELINE_DATA_SAMPLE_VIEW_H_
+
+#include <stdexcept>
+#include <type_traits>
+#include <utility>
+#include <vector>
+#include "dali/core/tensor_shape.h"
+#include "dali/pipeline/data/types.h"
+
+namespace dali {
+
+/**
+ * @brief SampleView - non-owning view type, keeping the pointer, shape and runtime type information
+ * of specific sample.
+ *
+ * This is a type corresponding to Tensor<Backend> and it's main goal is to become the return type
+ * for the TensorVector/TensorList::operator[]. It allows to access the data via the
+ * convenient `view<T, ndim>(SampleView)` conversion to TensorView, but doesn't break the batch
+ * object encapsulation and doesn't allow to adjust the allocation.
+ */
+template <typename Backend>
+class SampleView {
+ public:
+  /**
+   * @name Get the underlying pointer to data
+   */
+  // @{
+  void *data() {
+    return data_;
+  }
+
+  const void *data() const {
+    return data_;
+  }
+
+  const void *cdata() const {
+    return data_;
+  }
```

Comment:
If this type is to stay, I'd suggest that the API be consistent with existing buffer types.

Reply:
I would really like to introduce some more std-like naming here, but maybe as a next step.